sdk/azidentity/interactive_browser_credential.go

```diff
@@ -106,36 +109,49 @@ func (c *InteractiveBrowserCredential) AuthenticationPolicy(options azcore.Authe
-var authCodeReceiver = func(authorityHost string, tenantID string, clientID string, redirectURI string, port int, scopes []string) (*interactiveConfig, error) {
-	return interactiveBrowserLogin(authorityHost, tenantID, clientID, redirectURI, port, scopes)
+var authCodeReceiver = func(authorityHost string, opts *InteractiveBrowserCredentialOptions, scopes []string) (*interactiveConfig, error) {
+	return interactiveBrowserLogin(authorityHost, opts, scopes)
 }
 
 // interactiveBrowserLogin opens an interactive browser with the specified tenant and client IDs provided then returns the authorization code
 // received or an error.
-func interactiveBrowserLogin(authorityHost string, tenantID string, clientID string, redirectURL string, port int, scopes []string) (*interactiveConfig, error) {
-	const authPathFormat = "%s/oauth2/v2.0/authorize?response_type=code&response_mode=query&client_id=%s&redirect_uri=%s&state=%s&scope=%s&prompt=select_account"
-	state := func() string {
-		rand.Seed(time.Now().Unix())
-		// generate a 20-char random alpha-numeric string
-		const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
-		buff := make([]byte, 20)
-		for i := range buff {
-			buff[i] = charset[rand.Intn(len(charset))]
-		}
-		return string(buff)
-	}()
+func interactiveBrowserLogin(authorityHost string, opts *InteractiveBrowserCredentialOptions, scopes []string) (*interactiveConfig, error) {
 	// start local redirect server so login can call us back
 	rs := newServer()
+	state := uuid.New().String()
+	redirectURL := opts.RedirectURL
 	if redirectURL == "" {
-		redirectURL = rs.Start(state, port)
+		redirectURL = rs.Start(state, opts.Port)
 	}
 	defer rs.Stop()
 	u, err := url.Parse(authorityHost)
 	if err != nil {
 		return nil, err
 	}
-	authPath := fmt.Sprintf(authPathFormat, tenantID, clientID, redirectURL, state, strings.Join(scopes, " "))
-	u.Path = azcore.JoinPaths(u.Path, authPath)
+	values := url.Values{}
+	values.Add("response_type", "code")
+	values.Add("response_mode", "query")
+	values.Add("client_id", opts.ClientID)
+	values.Add("redirect_uri", redirectURL)
+	values.Add("state", state)
+	values.Add("scope", strings.Join(scopes, " "))
+	values.Add("prompt", "select_account")
+	cv := ""
+	if opts.DisablePKCE == false {
+		// the code verifier is a random 32-byte sequence that's been base-64 encoded without padding.
+		// it's used to prevent MitM attacks during auth code flow, see https://tools.ietf.org/html/rfc7636
+		b := make([]byte, 32, 32)
+		if _, err := rand.Read(b); err != nil {
+			return nil, err
+		}
+		cv = base64.RawURLEncoding.EncodeToString(b)
+		// for PKCE, create a hash of the code verifier
+		cvh := sha256.Sum256([]byte(cv))
+		values.Add("code_challenge", base64.RawURLEncoding.EncodeToString(cvh[:]))
+		values.Add("code_challenge_method", "S256")
+	}
+	u.Path = azcore.JoinPaths(u.Path, opts.TenantID, "oauth2/v2.0/authorize")
```

Comment:
Wondering if `"oauth2/v2.0/authorize"` should be in a const?

Reply:
Fixed